### DISCUSSION TITLE
Difficult setting up SMTP

### DISCUSSION THREAD

**Discussion by ecs0set** [dani-garcia/vaultwarden#5223] [General]
Hi,

I'm trying to setup Vaultwarden on a new Asustor device and need to setup SMTP.  Asustor has a mail server (based on postfix) which I have configured, and I've set Vaultwarden Admin page with the correct details for this SMTP server.  SMTP from other apps appears to be working correctly.

The SMTP connection is failing, I believe due to a TLS issue.  I believe Postfix is configured to require SSL 1.3 and it looks like Vaultwarden is not using that protocol (I believe it's using the same LetsEncrypt certificate as the other apps).  Rather than changing postfix to allow insecure SSL protocols, I'd like to bring Vaultwarden up to SSL 1.3

Please can you advise how to configure the SSL protocol used by Vaultwarden?  Or alternatively, is there anything else that is likely to be causing this?

Thanks!

**Comment by BlackDex** [comment]
Enable `SMTP_DEBUG=true` and check the logs.
Vaultwarden supports TLS1.3 without issues as far as i know

**Reply by paolobarbolini** [reply]
FWIW I don't know if it still applies since I don't use it but there's https://github.com/sfackler/rust-native-tls/issues/140

**Reply by BlackDex** [reply]
We do use native now. But without logs seeing what actually is happening I'm not sure switching to rust-tls will help then.

We still need OpenSSL unfortunately because of PostgreSQL.
But other libraries could be changed

**Comment by ecs0set** [comment]
I've enabled the SMTP debug...  I think.  I'm adding it to config.json rather than adjusting the Docker environment, is that sufficient?

**bitwarden_rs.log:**
[2024-11-23 16:19:23.027][vaultwarden::mail][ERROR] SMTP error: Connection error: Connection error: error:0A00042E:SSL routines:ssl3_read_bytes:tlsv1 alert protocol version:../ssl/record/rec_layer_s3.c:1605:SSL alert number 70

**maillog:**
Nov 23 16:19:23 <myservername> postfix/smtps/smtpd[2611]: connect from unknown[172.17.0.2]
Nov 23 16:19:23 <myservername>postfix/smtps/smtpd[2611]: SSL_accept error from unknown[172.17.0.2]: -1
Nov 23 16:19:23 <myservername>postfix/smtps/smtpd[2611]: warning: TLS library problem: error:14209102:SSL routines:tls_early_post_process_client_hello:unsupported protocol:../../../source/openssl-1.1.1w/ssl/statem/statem_srvr.c:1686:
Nov 23 16:19:23 <myservername>postfix/smtps/smtpd[2611]: lost connection after CONNECT from unknown[172.17.0.2]
Nov 23 16:19:23 <myservername>postfix/smtps/smtpd[2611]: disconnect from unknown[172.17.0.2] commands=0/0

**Vaultwarden config:**
  "_enable_smtp": true,
  "use_sendmail": false,
  "smtp_host": "smtp.<mydnsdomain>",
  "smtp_security": "force_tls",
  "smtp_port": 465,
  "smtp_from": "vaultwarden@<mydnsdomain>",
  "smtp_from_name": "Vaultwarden",
  "smtp_username": "smtp@<mydnsdomain>",
  "smtp_password": "<my password in plain text?!>",
  "smtp_auth_mechanism": "Plain",
  "smtp_timeout": 15,
  "smtp_embed_images": true,
  "smtp_accept_invalid_certs": false,
  "smtp_accept_invalid_hostnames": false,
  "_enable_email_2fa": false,
  "email_token_size": 6,
  "email_expiration_time": 600,
  "email_attempts_limit": 3,
  "email_2fa_enforce_on_verified_invite": false,
  "email_2fa_auto_fallback": false

**Postfix config (extracts):**
smtpd_recipient_restrictions = permit_mynetworks, permit_sasl_authenticated, reject_unknown_recipient_domain, reject_unauth_pipelining, reject_unauth_destination
smtpd_relay_restrictions = permit_mynetworks, permit_sasl_authenticated, defer_unauth_destination
smtpd_sasl_auth_enable = yes
smtpd_sasl_security_options = noanonymous
smtpd_tls_auth_only = yes
smtpd_tls_cert_file = /usr/builtin/etc/certificate/ssl.crt
smtpd_tls_key_file = /usr/builtin/etc/certificate/ssl.key
smtpd_tls_mandatory_ciphers = medium
smtpd_tls_mandatory_protocols = SSLv3, TLSv1
smtpd_tls_received_header = yes
smtpd_use_tls = yes
tls_random_source = dev:/dev/urandom
smtpd_sasl_authenticated_header = yes

I note there were issues with SMTP servers using LetsEncrypt certificates previously, seemingly resolved by an update to Bitwarden .NET Core?  Did the same issue affect Vaultwarden?
https://github.com/bitwarden/server/issues/370
https://github.com/bitwarden/server/issues/457
https://github.com/bitwarden/server/issues/1176

**Reply by ecs0set** [reply]
Ah, I've sussed it.  It was the mandatory protocols NOT including TLSv1.2 and TLSv1.3.  Poor defaults by Asustor there!

Now fixed, thanks!
